pkg/network/protocols/http/shared_libraries.go

```diff
@@ -35,91 +37,189 @@ func (l *libPath) Bytes() []byte {
+	binary.LittleEndian.PutUint64(b[8:], p.inode)
+	m := murmur3.Sum64(b)
+	binary.LittleEndian.PutUint64(b, m)
+	return base64.StdEncoding.EncodeToString(b[:8])
+}
+
+// path must be an absolute path
+func newPathIdentifier(path string) (pi pathIdentifier, err error) {
+	if len(path) < 1 || path[0] != '/' {
+		return pi, fmt.Errorf("invalid path %s", path)
+	}
+	info, err := os.Stat(path)
+	if err != nil {
+		return pi, err
+	}
 
-// check if a process is still alive, if not unregister his hook
-// when attaching a uprobe the kernel modifiy the elf and lock the file, so the filesystem can't be unmounted
-const soCheckProcessAliveInterval = 10 * time.Second
+	stat, ok := info.Sys().(*syscall.Stat_t)
+	if !ok {
+		return pi, fmt.Errorf("invalid file stat")
+	}
+
+	return pathIdentifier{
+		dev:   stat.Dev,
+		inode: stat.Ino,
+	}, nil
+}
 
 type soRule struct {
 	re           *regexp.Regexp
-	registerCB   func(string) error
-	unregisterCB func(string) error
+	registerCB   func(id pathIdentifier, root string, path string) error
+	unregisterCB func(id pathIdentifier) error
 }
 
 // soWatcher provides a way to tie callback functions to the lifecycle of shared libraries
 type soWatcher struct {
-	procRoot   string
-	hostMount  string
-	all        *regexp.Regexp
-	rules      []soRule
-	loadEvents *ddebpf.PerfHandler
-	registry   *soRegistry
+	procRoot       string
+	all            *regexp.Regexp
+	rules          []soRule
+	loadEvents     *ddebpf.PerfHandler
+	processMonitor *monitor.ProcessMonitor
+	registry       *soRegistry
 }
 
-type seenKey struct {
-	pid, path string
+type soRegistry struct {
+	m     sync.Mutex
+	byID  map[pathIdentifier]*soRegistration
+	byPID map[uint32]*soRegistration
+
+	// if we can't register a uprobe we don't try more than once
+	blacklistByID map[pathIdentifier]struct{}
 }
 
-func newSOWatcher(procRoot string, perfHandler *ddebpf.PerfHandler, rules ...soRule) *soWatcher {
+func newSOWatcher(perfHandler *ddebpf.PerfHandler, rules ...soRule) *soWatcher {
 	allFilters := make([]string, len(rules))
 	for i, r := range rules {
 		allFilters[i] = r.re.String()
 	}
 
 	all := regexp.MustCompile(fmt.Sprintf("(%s)", strings.Join(allFilters, "|")))
 	return &soWatcher{
-		procRoot:   procRoot,
-		hostMount:  os.Getenv("HOST_ROOT"),
-		all:        all,
-		rules:      rules,
-		loadEvents: perfHandler,
+		procRoot:       util.GetProcRoot(),
+		all:            all,
+		rules:          rules,
+		loadEvents:     perfHandler,
+		processMonitor: monitor.GetProcessMonitor(),
 		registry: &soRegistry{
-			procRoot: procRoot,
-			byPath:   make(map[string]*soRegistration),
-			byInode:  make(map[uint64]*soRegistration),
+			byID:          make(map[pathIdentifier]*soRegistration),
+			byPID:         make(map[uint32]*soRegistration),
+			blacklistByID: make(map[pathIdentifier]struct{}),
 		},
 	}
 }
 
+type soRegistration struct {
+	pathID pathIdentifier
+	//mTime  uint64 TODO(nplanel) reload uprobes if the file changed (dnf upgrade)
```

Comment:
is it still relevant?

Reply:
yes we don't support upgrade
I created a jira ticket https://datadoghq.atlassian.net/browse/USMO-172?atlOrigin=eyJpIjoiZGUzMGUyY2MzMWRkNDFjZGE3YjBhZWJkM2YyMDU3YzciLCJwIjoiaiJ9